chainerx_cc/chainerx/routines/math.cc

```diff
@@ -1005,6 +1005,42 @@ Array Arctan(const Array& x) {
     return out;
 }
 
+Array Arctan2(const Array& x1, const Array& x2) {
+    Dtype out_dtype = GetMathResultDtype(GetArithmeticResultDtype(x1, x2));
+    Array x1_cast = x1.dtype() != out_dtype ? x1.AsType(out_dtype) : x1;
```

Comment:
Avoid redundant copy.
```suggestion
    const Array& x1_cast = x1.dtype() != out_dtype ? x1.AsType(out_dtype) : x1;
```

Reply:
Could you move these casts to device implementation?

Reply:
The device already casts them, so removed it from here, and backward takes care of the type of `gradient Array` similar to `AddImpl`